File: sat_gt/sat_gt/page/fel_importer/fel_importer.py
"""Backend methods for the FEL importer test page."""

from __future__ import annotations

import base64
from calendar import monthrange
from datetime import date
from decimal import Decimal

import frappe

from sat_gt.fel.erpnext import factura_electronica_is_installed
from sat_gt.fel.matching import find_purchase_invoice_matches
from sat_gt.fel.parser import FELDocument, parse_fel_xml


@frappe.whitelist()
def parse_uploaded_xml(filename: str, content: str) -> dict:
	"""Parse one base64-encoded XML and return data for the test page."""
	try:
		raw_xml = base64.b64decode(content, validate=True)
		document = parse_fel_xml(raw_xml)
	except Exception as exc:
		return {"filename": filename, "ok": False, "error": str(exc)}

	result = {"filename": filename, "ok": True, "document": _serialize_document(document)}
	if factura_electronica_is_installed() and not document.is_credit_note:
		result["supplier"] = _find_supplier(document)
		result["matches"] = _find_matches(document)
	else:
		result["matches"] = []
	return result


@frappe.whitelist()
def get_unmatched_purchase_invoices(month: str, total: str | None = None, currency: str | None = None) -> list[dict]:
	"""Return Purchase Invoices from a month without an assigned FEL UUID."""
	_require_fel_app()
	year, month_number = (int(value) for value in month.split("-"))
	start = date(year, month_number, 1)
	end = date(year, month_number, monthrange(year, month_number)[1])
	filters = {
		"posting_date": ["between", [start.isoformat(), end.isoformat()]],
		"numero_autorizacion_fel": ["in", ["", None]],
		"docstatus": ["!=", 2],
	}
	if currency:
		filters["currency"] = currency
	if total:
		filters["grand_total"] = ["between", [str(Decimal(total) - Decimal("0.01")), str(Decimal(total) + Decimal("0.01"))]]
	return frappe.get_all(
		"Purchase Invoice",
		filters=filters,
		fields=["name", "posting_date", "supplier", "supplier_name", "facelec_nit_fproveedor", "grand_total", "currency"],
		order_by="posting_date desc, name desc",
		limit_page_length=0,
	)


@frappe.whitelist()
def associate_uuid_manually(purchase_invoice: str, content: str, tolerance: str = "0.01") -> dict:
	"""Associate a selected existing invoice with the UUID from an XML."""
	_require_fel_app()
	document = parse_fel_xml(_decode_xml(content))
	invoice = frappe.get_doc("Purchase Invoice", purchase_invoice)
	if invoice.numero_autorizacion_fel:
		frappe.throw("La Purchase Invoice ya tiene un UUID asociado")
	if frappe.db.exists("Purchase Invoice", {"numero_autorizacion_fel": document.uuid}):
		frappe.throw("El UUID ya está asociado a otra Purchase Invoice")
	if invoice.currency != document.currency:
		frappe.throw("La moneda de la factura no coincide con la del XML")
	if abs(Decimal(str(invoice.grand_total)) - document.grand_total) > Decimal(tolerance):
		frappe.throw("El total de la factura no coincide con el XML")

	invoice.numero_autorizacion_fel = document.uuid
	invoice.serie_original_del_documento = document.authorization_series or ""
	invoice.save()
	return {"name": invoice.name, "uuid": document.uuid}


@frappe.whitelist()
def create_imported_suppliers(documents: str) -> dict:
	"""Create missing suppliers in the dedicated imported supplier group."""
	_require_fel_app()
	rows = frappe.parse_json(documents) if isinstance(documents, str) else documents
	group_name = "Proveedores Importados"
	if not frappe.db.exists("Supplier Group", group_name):
		frappe.get_doc({"doctype": "Supplier Group", "supplier_group_name": group_name}).insert()

	created = []
	existing = []
	errors = []
	for row in rows:
		try:
			document = parse_fel_xml(_decode_xml(row["content"]))
			supplier = frappe.db.get_value("Supplier", {"facelec_nit_proveedor": document.issuer_nit}, "name")
			if supplier:
				existing.append({"nit": document.issuer_nit, "name": supplier})
				continue
			supplier_doc = frappe.get_doc(
				{
					"doctype": "Supplier",
					"supplier_name": document.issuer_name or document.issuer_nit,
					"supplier_group": group_name,
					"supplier_type": "Company",
					"tax_id": document.issuer_nit,
					"facelec_nit_proveedor": document.issuer_nit,
				}
			).insert()
			created.append({"nit": document.issuer_nit, "name": supplier_doc.name})
		except Exception as exc:
			errors.append({"filename": row.get("filename", ""), "error": str(exc)})
	return {"group": group_name, "created": created, "existing": existing, "errors": errors}


@frappe.whitelist()
def create_purchase_invoice_draft(
	content: str,
	company: str,
	supplier: str | None = None,
	default_item: str | None = None,
	expense_account: str | None = None,
) -> dict:
	"""Create an unsubmitted Purchase Invoice draft from a FEL XML.

	The initial draft uses the XML line total as the gross rate. Tax account
	mapping remains explicit in the next accounting configuration step.
	"""
	_require_fel_app()
	document = parse_fel_xml(_decode_xml(content))
	if document.is_credit_note:
		frappe.throw("Las notas de crédito se implementarán como un flujo separado")
	if frappe.db.exists("Purchase Invoice", {"numero_autorizacion_fel": document.uuid}):
		frappe.throw("El UUID ya está asociado a una Purchase Invoice")

	matched_supplier = _find_supplier(document)
	if not matched_supplier:
		frappe.throw(
			f"No existe un proveedor con el NIT {document.issuer_nit}. "
			"Créalo primero usando Crear proveedores importados."
		)
	if supplier and supplier != matched_supplier["name"]:
		frappe.throw("El proveedor no coincide con el NIT del XML")
	supplier = matched_supplier["name"]
	if not default_item and not expense_account:
		frappe.throw("Selecciona un Item por defecto o una cuenta de gasto")

	invoice = frappe.get_doc(
		{
			"doctype": "Purchase Invoice",
			"company": company,
			"supplier": supplier,
			"posting_date": document.emission_datetime.date().isoformat(),
			"bill_no": document.uuid,
			"bill_date": document.emission_datetime.date().isoformat(),
			"currency": document.currency,
			"facelec_nit_fproveedor": document.issuer_nit,
			"numero_autorizacion_fel": document.uuid,
			"serie_original_del_documento": document.authorization_series or "",
		}
	)
	for item in document.items:
		row = {
			"description": item.description,
			"qty": item.quantity,
			"rate": item.total / item.quantity,
		}
		if default_item:
			row["item_code"] = default_item
		if expense_account:
			row["expense_account"] = expense_account
		invoice.append("items", row)
	invoice.insert()
	return {"name": invoice.name, "uuid": document.uuid, "warning": "Borrador creado con importes brutos; revisar impuestos antes de contabilizar."}


def _find_matches(document: FELDocument) -> list[dict]:
	"""Find candidates for display without modifying any ERPNext document."""
	invoices = frappe.get_all(
		"Purchase Invoice",
		filters={"docstatus": ["!=", 2]},
		fields=["name", "supplier", "supplier_name", "facelec_nit_fproveedor", "grand_total", "currency"],
		limit_page_length=0,
	)
	matches = find_purchase_invoice_matches(document, invoices)
	return [
		{
			"name": candidate.document.name,
			"supplier": candidate.document.supplier,
			"supplier_name": candidate.document.supplier_name,
			"grand_total": str(candidate.document.grand_total),
			"currency": candidate.document.currency,
		}
		for candidate in matches
	]


def _decode_xml(content: str) -> bytes:
	try:
		return base64.b64decode(content, validate=True)
	except Exception as exc:
		raise ValueError("Contenido XML inválido") from exc


def _find_supplier(document: FELDocument) -> dict | None:
	"""Find a supplier by FEL NIT, accepting common hyphen formatting."""
	variants = {document.issuer_nit, document.issuer_nit.replace("-", "")}
	for nit in variants:
		for field in ("facelec_nit_proveedor", "tax_id"):
			name = frappe.db.get_value("Supplier", {field: nit}, "name")
			if name:
				return frappe.db.get_value(
					"Supplier",
					name,
					["name", "supplier_name", "facelec_nit_proveedor"],
					as_dict=True,
				)
	return None


def _require_fel_app() -> None:
	if not factura_electronica_is_installed():
		frappe.throw("La app factura_electronica es requerida para esta operación")


def _serialize_document(document: FELDocument) -> dict:
	return {
		"uuid": document.uuid,
		"authorization_number": document.authorization_number,
		"authorization_series": document.authorization_series,
		"issuer_nit": document.issuer_nit,
		"issuer_name": document.issuer_name,
		"document_type": document.document_type,
		"emission_datetime": document.emission_datetime.isoformat(),
		"currency": document.currency,
		"grand_total": str(document.grand_total),
		"receiver_nit": document.receiver_nit,
		"receiver_name": document.receiver_name,
		"complements": list(document.complements),
		"taxes": [{"name": tax.name, "amount": str(tax.tax_amount)} for tax in document.taxes],
		"items": [
			{
				"line_number": item.line_number,
				"description": item.description,
				"quantity": str(item.quantity),
				"unit": item.unit,
				"unit_price": str(item.unit_price),
				"total": str(item.total),
			}
			for item in document.items
		],
	}

File: sat_gt/sat_gt/page/fel_importer/fel_importer.js
frappe.pages["fel-importer"].on_page_load = function (wrapper) {
	new SatGtFelImporter(wrapper);
};

class SatGtFelImporter {
	constructor(wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Importador FEL"),
			single_column: true,
		});
		this.render();
	}

	render() {
		this.page.main.html(`
			<div class="fel-importer">
				<div class="fel-importer-intro">
					<h3>${__("Probar XML FEL")}</h3>
					<p class="text-muted">${__("Selecciona uno o varios XML. Se parsearán y se buscarán coincidencias contra Purchase Invoice sin modificar documentos.")}</p>
				</div>
				<div class="form-group">
					<label>${__("Archivos XML")}</label>
					<input class="form-control fel-files" type="file" accept=".xml,text/xml" multiple>
				</div>
				<div class="form-row align-items-end mb-3">
					<div class="form-group col-sm-4 mb-0">
						<label>${__("Mes para asociación manual")}</label>
						<input class="form-control fel-month" type="month">
					</div>
					<div class="col-sm-8 text-muted small pb-2">${__("Busca Purchase Invoice sin UUID y permite asociarla por total.")}</div>
				</div>
				<div class="fel-actions"></div>
				<div class="fel-results"></div>
			</div>
		`);
		this.$files = this.page.main.find(".fel-files");
		this.$actions = this.page.main.find(".fel-actions");
		this.$results = this.page.main.find(".fel-results");
		this.$month = this.page.main.find(".fel-month");
		this.$files.on("change", () => this.showSelectedFiles());
	}

	showSelectedFiles() {
		const files = Array.from(this.$files[0].files || []);
		this.$actions.empty();
		this.$results.empty();
		if (!files.length) return;

		const button = $(`<button class="btn btn-primary">${__("Parsear archivos")}</button>`);
		button.on("click", () => this.parseFiles(files, button));
		this.$actions.append(button);
		this.$actions.append(`<span class="text-muted ml-3">${files.length} ${__("archivo(s) seleccionado(s)")}</span>`);
	}

	async parseFiles(files, button) {
		button.prop("disabled", true).text(__("Procesando..."));
		this.$results.html(`<div class="text-muted">${__("Analizando XML...")}</div>`);
		const results = [];
		for (const file of files) {
			try {
				const content = await this.readAsBase64(file);
				const response = await frappe.call({
					method: "sat_gt.sat_gt.page.fel_importer.fel_importer.parse_uploaded_xml",
					args: { filename: file.name, content },
				});
				const result = response.message;
				result.content = content;
				results.push(result);
			} catch (error) {
				results.push({ filename: file.name, ok: false, error: error.message || String(error) });
			}
		}
		this.results = results;
		this.renderResults(results);
		this.renderSupplierAction(results);
		button.prop("disabled", false).text(__("Parsear archivos"));
	}

	readAsBase64(file) {
		return new Promise((resolve, reject) => {
			const reader = new FileReader();
			reader.onload = () => resolve(reader.result.split(",")[1]);
			reader.onerror = () => reject(reader.error || new Error(__("No se pudo leer el archivo")));
			reader.readAsDataURL(file);
		});
	}

	renderResults(results) {
		this.$results.empty();
		results.forEach((result) => {
			this.$results.append(result.ok ? this.documentCard(result) : this.errorCard(result));
		});
		this.$results.find(".fel-find-month").on("click", (event) => {
			const result = results[Number($(event.currentTarget).data("result-index"))];
			this.findMonthlyInvoice(result);
		});
		this.$results.find(".fel-create-draft").on("click", (event) => {
			const result = results[Number($(event.currentTarget).data("result-index"))];
			this.createDraft(result);
		});
	}

	renderSupplierAction(results) {
		this.$actions.find(".fel-suppliers").remove();
		const documents = results.filter((result) => result.ok);
		if (!documents.length) return;
		const button = $(`<button class="btn btn-secondary ml-2 fel-suppliers">${__("Crear proveedores importados")}</button>`);
		button.on("click", () => this.createSuppliers(documents, button));
		this.$actions.append(button);
	}

	createSuppliers(results, button) {
		frappe.confirm(
			__("Se crearán proveedores faltantes en el grupo Proveedores Importados. ¿Continuar?"),
			() => {
				button.prop("disabled", true);
				frappe.call({
					method: "sat_gt.sat_gt.page.fel_importer.fel_importer.create_imported_suppliers",
					args: { documents: JSON.stringify(results.map((result) => ({ filename: result.filename, content: result.content }))) },
				}).then((response) => {
					const data = response.message;
					frappe.msgprint(__("Creados: {0}. Existentes: {1}. Errores: {2}.", [data.created.length, data.existing.length, data.errors.length]));
					button.prop("disabled", false);
				});
			},
		);
	}

	findMonthlyInvoice(result) {
		if (!this.$month.val()) {
			frappe.msgprint(__("Selecciona un mes para buscar la factura."));
			return;
		}
		frappe.call({
			method: "sat_gt.sat_gt.page.fel_importer.fel_importer.get_unmatched_purchase_invoices",
			args: { month: this.$month.val(), total: result.document.grand_total, currency: result.document.currency },
		}).then((response) => {
			const candidates = response.message || [];
			if (!candidates.length) {
				frappe.msgprint(__("No hay facturas sin UUID con ese total en el mes seleccionado."));
				return;
			}
			const options = candidates.map((candidate) => candidate.name).join("\n");
			frappe.prompt(
				[{ fieldname: "invoice", fieldtype: "Select", label: __("Purchase Invoice"), options }],
				(values) => this.associateInvoice(values.invoice, result),
				__("Asociar UUID"),
				__("Asociar"),
			);
		});
	}

	associateInvoice(invoice, result) {
		frappe.call({
			method: "sat_gt.sat_gt.page.fel_importer.fel_importer.associate_uuid_manually",
			args: { purchase_invoice: invoice, content: result.content },
		}).then(() => {
			frappe.msgprint(__("UUID asociado correctamente."));
		});
	}

	createDraft(result) {
		const taxFields = (result.document.taxes || []).map((tax) => ({
			fieldname: this.taxFieldName(tax.name),
			fieldtype: "Link",
			options: "Account",
			label: `${__("Cuenta")} ${tax.name} (${tax.amount})`,
			reqd: 1,
		}));
		frappe.prompt(
			[
				{ fieldname: "company", fieldtype: "Link", options: "Company", label: __("Compañía"), reqd: 1 },
				{ fieldname: "default_item", fieldtype: "Link", options: "Item", label: __("Item por defecto") },
				{ fieldname: "expense_account", fieldtype: "Data", label: __("Cuenta de gasto (si no usa Item)") },
				...taxFields,
			],
			(values) => {
				const taxAccounts = {};
				(result.document.taxes || []).forEach((tax) => {
					taxAccounts[tax.name] = values[this.taxFieldName(tax.name)];
				});
				frappe.call({
					method: "sat_gt.sat_gt.page.fel_importer.fel_importer.create_purchase_invoice_draft",
					args: { ...values, tax_accounts: JSON.stringify(taxAccounts), content: result.content },
				}).then((response) => {
					frappe.msgprint(`${__("Borrador creado")}: ${response.message.name}<br>${response.message.warning}`);
				});
			},
			__("Crear borrador"),
			__("Crear"),
		);
	}

	taxFieldName(name) {
		return `tax_account_${name.toLowerCase().replace(/[^a-z0-9]+/g, "_")}`;
	}

		documentCard(result) {
		const doc = result.document;
		const matches = result.matches || [];
		const matchHtml = matches.length
			? matches.map((match) => `<li><a href="/app/purchase-invoice/${encodeURIComponent(match.name)}">${frappe.utils.escape_html(match.name)}</a> · ${frappe.utils.escape_html(match.supplier_name || match.supplier || "")} · ${match.currency} ${match.grand_total}</li>`).join("")
			: `<li class="text-muted">${__("Factura no ingresada")}</li>`;
		const items = (doc.items || []).map((item) => `<tr><td>${item.line_number}</td><td>${frappe.utils.escape_html(item.description)}</td><td>${item.quantity} ${item.unit || ""}</td><td class="text-right">${item.total}</td></tr>`).join("");
		const taxes = (doc.taxes || []).map((tax) => `${frappe.utils.escape_html(tax.name)}: ${tax.amount}`).join(" · ") || __("Sin impuestos");

		return $(
			`<div class="card mb-3">
				<div class="card-header"><strong>${frappe.utils.escape_html(result.filename)}</strong><span class="indicator-pill green ml-2">${doc.document_type}</span>${result.already_imported ? `<span class="indicator-pill orange ml-2">${__("Ya importado")}</span>` : ""}</div>
				<div class="card-body">
					<div class="row">
						${this.field(__("UUID"), doc.uuid, "col-sm-6")}
						${this.field(__("Proveedor"), `${doc.issuer_name} (${doc.issuer_nit})`, "col-sm-6")}
						${this.field(__("Proveedor ERPNext"), result.supplier ? `${result.supplier.supplier_name} (${result.supplier.name})` : __("No existe proveedor con este NIT"), "col-sm-6")}
						${this.field(__("Total"), `${doc.currency} ${doc.grand_total}`, "col-sm-3")}
						${this.field(__("Impuestos XML"), taxes, "col-sm-5")}
						${this.field(__("Emisión"), doc.emission_datetime, "col-sm-5")}
						${this.field(__("Complementos"), (doc.complements || []).join(", ") || "-", "col-sm-4")}
					</div>
					<div class="fel-card-actions mb-3"><button class="btn btn-xs btn-secondary fel-find-month" data-result-index="${this.results ? this.results.indexOf(result) : 0}">${__("Buscar por mes y total")}</button><button class="btn btn-xs btn-primary ml-2 fel-create-draft" data-result-index="${this.results ? this.results.indexOf(result) : 0}">${__("Crear borrador")}</button></div>
					<h5>${__("Coincidencias")}</h5><ul>${matchHtml}</ul>
					<h5>${__("Líneas")}</h5>
					<table class="table table-bordered table-sm"><thead><tr><th>#</th><th>${__("Descripción")}</th><th>${__("Cantidad")}</th><th class="text-right">${__("Total")}</th></tr></thead><tbody>${items}</tbody></table>
				</div>
			</div>`
		);
	}

	field(label, value, className) {
		return `<div class="${className} mb-3"><div class="text-muted small">${label}</div><div>${frappe.utils.escape_html(String(value || "-"))}</div></div>`;
	}

	errorCard(result) {
		return $(`<div class="alert alert-danger"><strong>${frappe.utils.escape_html(result.filename)}</strong><br>${frappe.utils.escape_html(result.error)}</div>`);
	}
}
